rust-runtime/aws-smithy-http-server/src/plugin/check_health.rs

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::future::Ready;
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+
+use futures_util::Future;
+use hyper::{Body, Request, Response};
+use pin_project_lite::pin_project;
+use tower::{util::Oneshot, Layer, Service, ServiceExt};
+
+use crate::body::BoxBody;
+
+use super::Either;
+
+/// A [`tower::Layer`] used to apply [`CheckHealthService`].
+#[derive(Clone, Debug)]
+pub struct CheckHealthLayer<PingHandler> {
+    health_check_uri: &'static str,
+    ping_handler: PingHandler,
+}
+
+impl CheckHealthLayer<()> {
+    pub fn new<HandlerFuture: Future<Output = Response<BoxBody>>, H: Fn(Request<Body>) -> HandlerFuture>(
+        health_check_uri: &'static str,
+        ping_handler: H,
+    ) -> CheckHealthLayer<H> {
+        CheckHealthLayer {
+            health_check_uri,
+            ping_handler,
+        }
+    }
+}
+
+pub type DefaultHandler = fn(Request<Body>) -> Ready<Response<BoxBody>>;
+
+impl<S, H: Clone> Layer<S> for CheckHealthLayer<H> {
+    type Service = CheckHealthService<H, S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CheckHealthService {
+            inner,
+            layer: self.clone(),
+        }
+    }
+}
+
+/// A middleware [`Service`] responsible for handling health check requests.
+#[derive(Clone, Debug)]
+pub struct CheckHealthService<H, S> {
+    inner: S,
+    layer: CheckHealthLayer<H>,
+}
+
+pin_project! {
+    /// A future that converts `F` into a compatible `S::Future`.
+    pub struct MappedHandlerFuture<R, S, F> {
+        #[pin]
+        inner: F,
+        pd: PhantomData<fn(R) -> S>,
+    }
+}
+
+impl<R, S, F> MappedHandlerFuture<R, S, F> {
+    fn new(inner: F) -> MappedHandlerFuture<R, S, F> {
+        Self { inner, pd: PhantomData }
+    }
+}
+
+impl<R, S: Service<R>, F: Future<Output = S::Response>> Future for MappedHandlerFuture<R, S, F> {
+    type Output = Result<S::Response, S::Error>;
+
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.project().inner.poll(cx).map(Ok)
+    }
+}
```

Comment:
Is there a simpler way to do this? I basically just want to `map` `HandlerFuture` from a `Future<Output = Response<BoxBody>>` to a `Future<Output = Result<Response<BoxBody>, S::Error>>`.

I tried using [`futures_util::future::Map`](https://docs.rs/futures-util/latest/futures_util/future/struct.Map.html) but I ran into a compiler error when trying to write the type in line 92 as (expected function pointer but got function item):

```rust
type Future = Either<
  Map<HandlerFuture, fn(Response<BoxBody) -> Result<Response<BoxBody>, S::Error>>,
  Oneshot<S, Request<Body>>
>;
```